### ISSUE TITLE
Zip Export: Maximum number of SAP GUI sessions reached

### ISSUE THREAD

**Issue by abap-weasel** [abapGit/abapGit#4833]
Occasionally I get the error message "Maximum number of SAP GUI sessions reached" from Zip Export. 

![image](https://user-images.githubusercontent.com/47646214/121672207-d1084600-caaf-11eb-9169-1e73b4be6424.png)

I am then just restarting abapgit and normally the issue disappears. Wondering what is the imact from that error? Is export incomplete when the issue occurs?

**Comment by larshp** [comment]
abapGit runs the serialization in parallel, see https://github.com/abapGit/abapGit/blob/main/src/objects/core/zcl_abapgit_serialize.clas.abap

yes, I think the export is incomplete when this error occurs

**Comment by abap-weasel** [comment]
Ok, I think I roughly understand it.. Not sure if I will be able to pull this off, but what do you think about...

(1) saving relation of task id (sy-tabix from L554) and tadir record in an attribute.
(2) improving error handling in METHOD on_end_of_task to be able to check for failed task executions and add it to queue again. This would probably require to replace lt_tadir in serialize() with an attribute in order to be able to manipulate it.

Maybe it would also viable to gradually reduce mv_free by 1 when the issue appreas until it falls back to sequential mode (1 task)?

**Comment by mbtools** [comment]
This needs further investigation before we change anything. "GUI sessions" is front-end related but the serializer run on the backend. There's also a logic to determine number of free processes with a fallback to sequential processing. So I don't see the connection between error and the core serializer process here.

First, turn off "parallel processing" in settings. Is the error still reproducible?

Looks like you have some ECATT objects. Maybe the ECATT serializer class is the cause. Anyway, the best would be to determine if it's related to specific objects or object types. So, yes, to improving error handling in `on_end_of_task`. Function `z_abapgit_serialize_parallel` should pass the object (`ls_item`) to the error message. It's a bit tricky since it's limited in length.

**Comment by abap-weasel** [comment]
Yeah, issue only comes up for ECTC and ECTD type. It looks like I can reproduce the issue when I open as many SAPGUI sessions as possible manually during serialization.

**Comment by abap-weasel** [comment]
It looks like ECATT Framework itself uses STARTING NEW TASK to improve performance which probably causes that issue if a new task is started when all modes are already used from abapgit serialization. Maybe also PBT initialization (function SPBT_INITIALIZE) from abapgit causes side effects for calls initiated in ECATT ?

This is a search for STARTING NEW TASK in SECATT package (not sure yet what is actually used during abapgit serialization):

![image](https://user-images.githubusercontent.com/47646214/121702479-366d2e80-cad2-11eb-9cad-ca2fd0055a09.png)

**Comment by larshp** [comment]
@abap-weasel suggest doing 
![image](https://user-images.githubusercontent.com/5888506/140637903-e51e2909-3b13-4f4a-9b37-eb4c1ee0ed51.png)
and then before clicking ok, set a breakpoint at `CALL FUNCTION` statements

**Comment by mbtools** [comment]
no action
